<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#What-is-H2O?" data-toc-modified-id="What-is-H2O?-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>What is H2O?</a></span></li><li><span><a href="#The-H2O-Object-System" data-toc-modified-id="The-H2O-Object-System-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>The H2O Object System</a></span></li><li><span><a href="#How-to-use-H2O-from-Python" data-toc-modified-id="How-to-use-H2O-from-Python-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>How to use H2O from Python</a></span><ul class="toc-item"><li><span><a href="#Import-H2o-Python-library" data-toc-modified-id="Import-H2o-Python-library-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Import H2o Python library</a></span></li><li><span><a href="#Starting-H2O-and-Inspecting-the-Cluster" data-toc-modified-id="Starting-H2O-and-Inspecting-the-Cluster-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Starting H2O and Inspecting the Cluster</a></span><ul class="toc-item"><li><span><a href="#To-connect-to-an-established-H2O-cluster" data-toc-modified-id="To-connect-to-an-established-H2O-cluster-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>To connect to an established H2O cluster</a></span></li></ul></li><li><span><a href="#Cluster-Status-Summary" data-toc-modified-id="Cluster-Status-Summary-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Cluster Status Summary</a></span></li><li><span><a href="#Create-an-H2OFrame-object" data-toc-modified-id="Create-an-H2OFrame-object-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Create an H2OFrame object</a></span><ul class="toc-item"><li><span><a href="#To-create-an-H2OFrame-object-from-a-Python-tuple" data-toc-modified-id="To-create-an-H2OFrame-object-from-a-Python-tuple-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>To create an H2OFrame object from a Python tuple</a></span><ul class="toc-item"><li><span><a href="#Without-giving-the-destination-frame-name" data-toc-modified-id="Without-giving-the-destination-frame-name-3.4.1.1"><span class="toc-item-num">3.4.1.1&nbsp;&nbsp;</span>Without giving the destination frame name</a></span></li><li><span><a href="#By-giving-the-destination-frame-name" data-toc-modified-id="By-giving-the-destination-frame-name-3.4.1.2"><span class="toc-item-num">3.4.1.2&nbsp;&nbsp;</span>By giving the destination frame name</a></span></li></ul></li><li><span><a href="#View-the-H2OFrame" data-toc-modified-id="View-the-H2OFrame-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>View the H2OFrame</a></span></li></ul></li><li><span><a href="#Listing-Cluster-Contents" data-toc-modified-id="Listing-Cluster-Contents-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Listing Cluster Contents</a></span></li><li><span><a href="#Assign-a-new-name" data-toc-modified-id="Assign-a-new-name-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Assign a new name</a></span></li><li><span><a href="#Remove-object-from-h2o" data-toc-modified-id="Remove-object-from-h2o-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Remove object from h2o</a></span><ul class="toc-item"><li><span><a href="#Remove-a-specific-object" data-toc-modified-id="Remove-a-specific-object-3.7.1"><span class="toc-item-num">3.7.1&nbsp;&nbsp;</span>Remove a specific object</a></span></li><li><span><a href="#Remove-all-the-objects-from-h2o" data-toc-modified-id="Remove-all-the-objects-from-h2o-3.7.2"><span class="toc-item-num">3.7.2&nbsp;&nbsp;</span>Remove all the objects from h2o</a></span></li></ul></li><li><span><a href="#Create-an-H2OFrame-object" data-toc-modified-id="Create-an-H2OFrame-object-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Create an H2OFrame object</a></span><ul class="toc-item"><li><span><a href="#To-create-an-H2OFrame-object-from-Python-tuples." data-toc-modified-id="To-create-an-H2OFrame-object-from-Python-tuples.-3.8.1"><span class="toc-item-num">3.8.1&nbsp;&nbsp;</span>To create an H2OFrame object from Python tuples.</a></span></li><li><span><a href="#To-create-an-H2OFrame-object-from-a-Python-list:" data-toc-modified-id="To-create-an-H2OFrame-object-from-a-Python-list:-3.8.2"><span class="toc-item-num">3.8.2&nbsp;&nbsp;</span>To create an H2OFrame object from a Python list:</a></span></li><li><span><a href="#To-create-an-H2OFrame-object-from-collections.OrderedDict-or-a-Python-dict" data-toc-modified-id="To-create-an-H2OFrame-object-from-collections.OrderedDict-or-a-Python-dict-3.8.3"><span class="toc-item-num">3.8.3&nbsp;&nbsp;</span>To create an H2OFrame object from collections.OrderedDict or a Python dict</a></span></li><li><span><a href="#To-create-an-H2OFrame-object-from-a-Python-dict-and-specify-the-column-types" data-toc-modified-id="To-create-an-H2OFrame-object-from-a-Python-dict-and-specify-the-column-types-3.8.4"><span class="toc-item-num">3.8.4&nbsp;&nbsp;</span>To create an H2OFrame object from a Python dict and specify the column types</a></span></li><li><span><a href="#Converting-pandas-data-frame-to-h2o-object" data-toc-modified-id="Converting-pandas-data-frame-to-h2o-object-3.8.5"><span class="toc-item-num">3.8.5&nbsp;&nbsp;</span>Converting pandas data frame to h2o object</a></span><ul class="toc-item"><li><span><a href="#Create-Pandas-DataFrame" data-toc-modified-id="Create-Pandas-DataFrame-3.8.5.1"><span class="toc-item-num">3.8.5.1&nbsp;&nbsp;</span>Create Pandas DataFrame</a></span></li><li><span><a href="#Converting-the-pandas-data-frame-to-h2o-frame" data-toc-modified-id="Converting-the-pandas-data-frame-to-h2o-frame-3.8.5.2"><span class="toc-item-num">3.8.5.2&nbsp;&nbsp;</span>Converting the pandas data frame to h2o frame</a></span></li></ul></li><li><span><a href="#Converting-the-h2o-frame-to-pandas-data-frame" data-toc-modified-id="Converting-the-h2o-frame-to-pandas-data-frame-3.8.6"><span class="toc-item-num">3.8.6&nbsp;&nbsp;</span>Converting the h2o frame to pandas data frame</a></span></li><li><span><a href="#To-display-the-top-and-bottom-of-an-H2OFrame" data-toc-modified-id="To-display-the-top-and-bottom-of-an-H2OFrame-3.8.7"><span class="toc-item-num">3.8.7&nbsp;&nbsp;</span>To display the top and bottom of an H2OFrame</a></span><ul class="toc-item"><li><span><a href="#View-top-10-rows-of-the-H2OFrame" data-toc-modified-id="View-top-10-rows-of-the-H2OFrame-3.8.7.1"><span class="toc-item-num">3.8.7.1&nbsp;&nbsp;</span>View top 10 rows of the H2OFrame</a></span></li><li><span><a href="#View-bottom-5-rows-of-the-H2OFrame" data-toc-modified-id="View-bottom-5-rows-of-the-H2OFrame-3.8.7.2"><span class="toc-item-num">3.8.7.2&nbsp;&nbsp;</span>View bottom 5 rows of the H2OFrame</a></span></li><li><span><a href="#To-display-the-column-names" data-toc-modified-id="To-display-the-column-names-3.8.7.3"><span class="toc-item-num">3.8.7.3&nbsp;&nbsp;</span>To display the column names</a></span></li><li><span><a href="#To-display-summary-statistics-of-your-data" data-toc-modified-id="To-display-summary-statistics-of-your-data-3.8.7.4"><span class="toc-item-num">3.8.7.4&nbsp;&nbsp;</span>To display summary statistics of your data</a></span></li></ul></li></ul></li><li><span><a href="#Selection" data-toc-modified-id="Selection-3.9"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>Selection</a></span><ul class="toc-item"><li><span><a href="#To-select-a-single-column-by-name,-resulting-in-an-H2OFrame" data-toc-modified-id="To-select-a-single-column-by-name,-resulting-in-an-H2OFrame-3.9.1"><span class="toc-item-num">3.9.1&nbsp;&nbsp;</span>To select a single column by name, resulting in an H2OFrame</a></span></li><li><span><a href="#To-select-a-single-column-by-index,-resulting-in-an-H2OFrame" data-toc-modified-id="To-select-a-single-column-by-index,-resulting-in-an-H2OFrame-3.9.2"><span class="toc-item-num">3.9.2&nbsp;&nbsp;</span>To select a single column by index, resulting in an H2OFrame</a></span></li><li><span><a href="#To-select-multiple-columns-by-name,-resulting-in-an-H2OFrame" data-toc-modified-id="To-select-multiple-columns-by-name,-resulting-in-an-H2OFrame-3.9.3"><span class="toc-item-num">3.9.3&nbsp;&nbsp;</span>To select multiple columns by name, resulting in an H2OFrame</a></span></li><li><span><a href="#To-select-multiple-columns-by-index,-resulting-in-an-H2OFrame:" data-toc-modified-id="To-select-multiple-columns-by-index,-resulting-in-an-H2OFrame:-3.9.4"><span class="toc-item-num">3.9.4&nbsp;&nbsp;</span>To select multiple columns by index, resulting in an H2OFrame:</a></span></li><li><span><a href="#To-select-multiple-rows-by-slicing,-resulting-in-an-H2OFrame:" data-toc-modified-id="To-select-multiple-rows-by-slicing,-resulting-in-an-H2OFrame:-3.9.5"><span class="toc-item-num">3.9.5&nbsp;&nbsp;</span>To select multiple rows by slicing, resulting in an H2OFrame:</a></span></li><li><span><a href="#To-select-rows-based-on-specic-criteria,-use-Boolean-masking" data-toc-modified-id="To-select-rows-based-on-specic-criteria,-use-Boolean-masking-3.9.6"><span class="toc-item-num">3.9.6&nbsp;&nbsp;</span>To select rows based on specic criteria, use Boolean masking</a></span></li></ul></li><li><span><a href="#Missing-Data" data-toc-modified-id="Missing-Data-3.10"><span class="toc-item-num">3.10&nbsp;&nbsp;</span>Missing Data</a></span><ul class="toc-item"><li><span><a href="#To-create-an-H2OFrame-from-Python-with-missing-elements" data-toc-modified-id="To-create-an-H2OFrame-from-Python-with-missing-elements-3.10.1"><span class="toc-item-num">3.10.1&nbsp;&nbsp;</span>To create an H2OFrame from Python with missing elements</a></span></li><li><span><a href="#To-determine-which-rows-are-missing-data-for-a-given-column-(`1'-indicates-missing):" data-toc-modified-id="To-determine-which-rows-are-missing-data-for-a-given-column-(`1'-indicates-missing):-3.10.2"><span class="toc-item-num">3.10.2&nbsp;&nbsp;</span>To determine which rows are missing data for a given column (`1' indicates missing):</a></span></li><li><span><a href="#To-determine-the-location-of-all-missing-data-in-an-H2OFrame" data-toc-modified-id="To-determine-the-location-of-all-missing-data-in-an-H2OFrame-3.10.3"><span class="toc-item-num">3.10.3&nbsp;&nbsp;</span>To determine the location of all missing data in an H2OFrame</a></span></li><li><span><a href="#To-change-all-missing-values-in-a-column-to-a-diferent-value:" data-toc-modified-id="To-change-all-missing-values-in-a-column-to-a-diferent-value:-3.10.4"><span class="toc-item-num">3.10.4&nbsp;&nbsp;</span>To change all missing values in a column to a diferent value:</a></span></li><li><span><a href="#Operations" data-toc-modified-id="Operations-3.10.5"><span class="toc-item-num">3.10.5&nbsp;&nbsp;</span>Operations</a></span></li><li><span><a href="#Handling-different-types-null-values" data-toc-modified-id="Handling-different-types-null-values-3.10.6"><span class="toc-item-num">3.10.6&nbsp;&nbsp;</span>Handling different types null values</a></span></li><li><span><a href="#Checking-for-number-null-values-at-column-level" data-toc-modified-id="Checking-for-number-null-values-at-column-level-3.10.7"><span class="toc-item-num">3.10.7&nbsp;&nbsp;</span>Checking for number null values at column level</a></span></li><li><span><a href="#Handling-missing-values" data-toc-modified-id="Handling-missing-values-3.10.8"><span class="toc-item-num">3.10.8&nbsp;&nbsp;</span>Handling missing values</a></span><ul class="toc-item"><li><span><a href="#Imputing-the-data-with-central-tendencies" data-toc-modified-id="Imputing-the-data-with-central-tendencies-3.10.8.1"><span class="toc-item-num">3.10.8.1&nbsp;&nbsp;</span>Imputing the data with central tendencies</a></span></li></ul></li></ul></li><li><span><a href="#apply-function" data-toc-modified-id="apply-function-3.11"><span class="toc-item-num">3.11&nbsp;&nbsp;</span>apply function</a></span></li><li><span><a href="#Merging" data-toc-modified-id="Merging-3.12"><span class="toc-item-num">3.12&nbsp;&nbsp;</span>Merging</a></span></li><li><span><a href="#Grouping" data-toc-modified-id="Grouping-3.13"><span class="toc-item-num">3.13&nbsp;&nbsp;</span>Grouping</a></span></li><li><span><a href="#Using-Date-and-Time-Data" data-toc-modified-id="Using-Date-and-Time-Data-3.14"><span class="toc-item-num">3.14&nbsp;&nbsp;</span>Using Date and Time Data</a></span></li><li><span><a href="#Categoricals" data-toc-modified-id="Categoricals-3.15"><span class="toc-item-num">3.15&nbsp;&nbsp;</span>Categoricals</a></span></li><li><span><a href="#Loading-and-Saving-Data" data-toc-modified-id="Loading-and-Saving-Data-3.16"><span class="toc-item-num">3.16&nbsp;&nbsp;</span>Loading and Saving Data</a></span><ul class="toc-item"><li><span><a href="#To-load-data-from-the-same-machine-running-H2O:" data-toc-modified-id="To-load-data-from-the-same-machine-running-H2O:-3.16.1"><span class="toc-item-num">3.16.1&nbsp;&nbsp;</span>To load data from the same machine running H2O:</a></span></li><li><span><a href="#To-load-data-from-the-machine(s)-running-H2O-to-the-machine-running-Python:" data-toc-modified-id="To-load-data-from-the-machine(s)-running-H2O-to-the-machine-running-Python:-3.16.2"><span class="toc-item-num">3.16.2&nbsp;&nbsp;</span>To load data from the machine(s) running H2O to the machine running Python:</a></span></li><li><span><a href="#To-save-an-H2OFrame-on-the-machine-running-H2O:" data-toc-modified-id="To-save-an-H2OFrame-on-the-machine-running-H2O:-3.16.3"><span class="toc-item-num">3.16.3&nbsp;&nbsp;</span>To save an H2OFrame on the machine running H2O:</a></span></li><li><span><a href="#To-save-an-H2OFrame-on-the-machine-running-Python:" data-toc-modified-id="To-save-an-H2OFrame-on-the-machine-running-Python:-3.16.4"><span class="toc-item-num">3.16.4&nbsp;&nbsp;</span>To save an H2OFrame on the machine running Python:</a></span><ul class="toc-item"><li><span><a href="#upload_file" data-toc-modified-id="upload_file-3.16.4.1"><span class="toc-item-num">3.16.4.1&nbsp;&nbsp;</span>upload_file</a></span></li><li><span><a href="#Cannot-load-the-data-from-extrenal-path.-Filesneeds-to-be-present-in-local-machine" data-toc-modified-id="Cannot-load-the-data-from-extrenal-path.-Filesneeds-to-be-present-in-local-machine-3.16.4.2"><span class="toc-item-num">3.16.4.2&nbsp;&nbsp;</span>Cannot load the data from extrenal path. Filesneeds to be present in local machine</a></span></li></ul></li></ul></li><li><span><a href="#Shutdwon-the-h2o-cluster" data-toc-modified-id="Shutdwon-the-h2o-cluster-3.17"><span class="toc-item-num">3.17&nbsp;&nbsp;</span>Shutdwon the h2o cluster</a></span></li></ul></li></ul></div>

# What is H2O?

H2O is a Java-based software for data modeling and general computing. The H2O software is many things, but the primary purpose of H2O is as a distributed (many machines), parallel (many CPUs), in memory (several hundred GBs Xmx) processing engine.

There are two levels of parallelism:

* Within node
* Across (or between) nodes

The goal of H2O is to allow simple __horizontal scaling__ to a given problem in order to produce a solution faster. The conceptual paradigm MapReduce (AKA “divide and conquer and combine”), along with a good concurrent application structure, (c.f. jsr166y and NonBlockingHashMap) enable this type of scaling in H2O.

For application developers and data scientists, the gritty details of thread-safety, algorithm parallelism, and node coherence on a network are concealed by simple-to-use REST calls that are all documented. In addition, H2O is an open-source project under the Apache v2 licence. All of the source code is on github.

# The H2O Object System

H2O uses a distributed key-value store (the “DKV”) that contains pointers to the various objects of the H2O ecosystem. Some shared objects are mutable by the client; some shared objects are read-only by the client, but are mutable by H2O (e.g. a model being constructed will change over time); and actions by the client may have side-effects on other clients (multi-tenancy is not a supported model of use, but it is possible for multiple clients to attach to a single H2O cluster).

Briefly, these objects are:

__Key__: A key is an entry in the DKV that maps to an object in H2O.

__Frame__: A Frame is a collection of Vec objects. It is a 2D array of elements.

__Vec__: A Vec is a collection of Chunk objects. It is a 1D array of elements.

__Chunk__: A Chunk holds a fraction of the BigData. It is a 1D array of elements.

__ModelMetrics__: A collection of metrics for a given category of model.

__Model__: A model is an immutable object having predict and metrics methods.

__Job__: A Job is a non-blocking task that performs a finite amount of work.

Many of these objects have no meaning to a Python end-user, but to make sense of the objects available in this module it is helpful to understand how these objects map to objects in the JVM. After all, this module is an interface that allows the manipulation of a distributed system.

# How to use H2O from Python

## Import H2o Python library

In [1]:
import h2o

## Starting H2O and Inspecting the Cluster

Every new python session begins by initializing a connection between the python client and the H2O cluster. 

In [2]:
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  ------------------------------------------------------------------
H2O_cluster_uptime:         06 secs
H2O_cluster_timezone:       Asia/Kolkata
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.30.1.2
H2O_cluster_version_age:    26 days
H2O_cluster_name:           H2O_from_python_jeevan_nrnp5c
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.887 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.8.5 final
--------------------------  ------------------------------------------------------------------

By default, this will attempt to discover an H2O at localhost:54321. Note that If it fails to find a running H2O instance at this address, it will seek out an h2o jar at several possible locations. If no jar is found, then an H2OStartupError will be raised

### To connect to an established H2O cluster

## Cluster Status Summary

After making a successful connection, you can obtain a high-level summary of the cluster status:

In [3]:
h2o.cluster().show_status()

--------------------------  ------------------------------------------------------------------
H2O_cluster_uptime:         06 secs
H2O_cluster_timezone:       Asia/Kolkata
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.30.1.2
H2O_cluster_version_age:    26 days
H2O_cluster_name:           H2O_from_python_jeevan_nrnp5c
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.887 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
H2O_API_Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, TargetEncoder, Core V4
Python_version:             3.8.5 final
--------------------------  ------------------------------------------------------------------

## Create an H2OFrame object 

### To create an H2OFrame object from a Python tuple

####  Without giving the destination frame name

In [4]:
df = h2o.H2OFrame(((1, 2, 3), ('a', 'b', 'c'), (0.1, 0.2, 0.3)))

Parse progress: |█████████████████████████████████████████████████████████| 100%


#### By giving the destination frame name

In [5]:
dftmp = h2o.H2OFrame(((1, 2, 3), ('a','b','c'), (0.1, 0.2, 0.3)), destination_frame = "dftmp.hex")

Parse progress: |█████████████████████████████████████████████████████████| 100%


### View the H2OFrame

In [6]:
df

C1 C2 C3 
 
 
 1 2 3 
 nan nan nan 
 0.1 0.2 0.3

In [7]:
dftmp

C1 C2 C3 
 
 
 1 2 3 
 nan nan nan 
 0.1 0.2 0.3

## Listing Cluster Contents

To list the current contents of the H2O cluster, you can use the h2o.ls command

In [8]:
h2o.ls()

key
0  Key_Frame__upload_a31cca13afff8e0f5d3b19cfdf14...
1                                          dftmp.hex
2                              ecg_discord_train.hex
3  http://h2o-public-test-data.s3.amazonaws.com/s...

There are models, data, and model metrics all floating around in the DKV.

## Assign a new name
    To destination_frame using python reference.

In [9]:
df = h2o.assign(data = df, xid = "df.hex" )

In [10]:
h2o.ls()

key
0  Key_Frame__upload_a31cca13afff8e0f5d3b19cfdf14...
1                                             df.hex
2                                          dftmp.hex
3                               ecg_discord_test.hex
4                              ecg_discord_train.hex
5  http://h2o-public-test-data.s3.amazonaws.com/s...

_Observation_: Old name is not replaced with new, one more alias is created for the same destination_frame

## Remove object from h2o

### Remove a specific object

In [11]:
h2o.remove("dftmp.hex")

Confirm by list the same using ls

In [12]:
h2o.ls()

key
0  Key_Frame__upload_a31cca13afff8e0f5d3b19cfdf14...
1                                             df.hex
2                               ecg_discord_test.hex
3                              ecg_discord_train.hex
4  http://h2o-public-test-data.s3.amazonaws.com/s...

### Remove all the objects from h2o

In [13]:
h2o.remove_all()

In [14]:
h2o.ls()

Empty DataFrame
Columns: [key]
Index: []

## Create an H2OFrame object 

### To create an H2OFrame object from Python tuples.

    Without giving the destination frame name

In [15]:
df = h2o.H2OFrame(((1, 2, 3), ('a', 'b', 'c'), (0.1, 0.2, 0.3)))

Parse progress: |█████████████████████████████████████████████████████████| 100%


View the H2OFrame

In [16]:
df

C1 C2 C3 
 
 
 1 2 3 
 nan nan nan 
 0.1 0.2 0.3

Observation: 
    
    Lists treated as rows of the table

In [17]:
df = h2o.H2OFrame(((1, 'a', 0.1), (2, 'b', 0.2),(3, 'c', 0.3)))

Parse progress: |█████████████████████████████████████████████████████████| 100%


View H2OFrame

In [18]:
df

C1 C2 C3 
 
 
 1 a 0.1 
 2 b 0.2 
 3 c 0.3

### To create an H2OFrame object from a Python list:

In [19]:
df = h2o.H2OFrame([[1, 2, 3], ['a', 'b', 'c'], [0.1, 0.2, 0.3]])

Parse progress: |█████████████████████████████████████████████████████████| 100%


View H2OFrame

In [20]:
df

C1 C2 C3 
 
 
 1 2 3 
 nan nan nan 
 0.1 0.2 0.3

In [21]:
df = h2o.H2OFrame([[1, 'a', 0.1], [2, 'b', 0.2], [3, 'c', 0.3]])

Parse progress: |█████████████████████████████████████████████████████████| 100%


View H2OFrame

In [22]:
df

C1 C2 C3 
 
 
 1 a 0.1 
 2 b 0.2 
 3 c 0.3

### To create an H2OFrame object from collections.OrderedDict or a Python dict

In [23]:
df = h2o.H2OFrame({'A': [1, 2, 3], 'B': ['a', 'b', 'c'], 'C': [0.1, 0.2, 0.3]})

Parse progress: |█████████████████████████████████████████████████████████| 100%


View the H2OFrame

In [24]:
df

A B C 
 
 
 1 a 0.1 
 2 b 0.2 
 3 c 0.3

To display the column types:

In [25]:
df.types

{'A': 'int', 'B': 'string', 'C': 'real'}

### To create an H2OFrame object from a Python dict and specify the column types

In [26]:
df2 = h2o.H2OFrame.from_python({"A": [1, 2, 3], 
                                "B": ["a", "a", "b"],
                                "C": ["hello", "all", "world"],
                                "D": ["12SEP2020:11:00:00", "13SEP2020:12:00:00", "14SEP2020:13:00:00"]},
                                column_types=["numeric", "enum", "string","time"])

Parse progress: |█████████████████████████████████████████████████████████| 100%


To display the column types:

In [27]:
df2.types

{'A': 'int', 'B': 'enum', 'C': 'string', 'D': 'time'}

View the H2OFrame

In [28]:
df2

A B C D 
 
 
 1 a hello 2020-09-12 11:00:00 
 2 a all 2020-09-13 12:00:00 
 3 b world 2020-09-14 13:00:00

### Converting pandas data frame to h2o object

#### Create Pandas DataFrame

In [29]:
import pandas as pd

# Python data frame
df_pandas = pd.DataFrame({'A': [1, 2, 3], 'B': ['a', 'b', 'c']})

# To check the aata type of object
type(df_pandas)

pandas.core.frame.DataFrame

In [30]:
df_pandas

A  B
0  1  a
1  2  b
2  3  c

#### Converting the pandas data frame to h2o frame

    dataframe.values returns the array list.

In [31]:
df = h2o.H2OFrame.from_python(df_pandas, column_names=list("AB"))

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [32]:
df

A B 
 
 
 1 a 
 2 b 
 3 c

In [33]:
type(df)

h2o.frame.H2OFrame

Observation: In future version H2o doesn't take Padans data frame directly. Either use  as_matrix() or take numpy arrary from pandas and pass it to H2o

In [34]:
df = h2o.H2OFrame.from_python(df_pandas.values, column_names=list("AB"))

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [35]:
df

A B 
 
 
 1 a 
 2 b 
 3 c

Just for understanding

In [36]:
list("AB")

['A', 'B']

Checking the data type of the object

In [37]:
type(df)

h2o.frame.H2OFrame

### Converting the h2o frame to pandas data frame

In [38]:
df_pand = h2o.H2OFrame.as_data_frame(df)

In [39]:
type(df_pand)

pandas.core.frame.DataFrame

In [40]:
df_pand

A  B
0  1  a
1  2  b
2  3  c

In [41]:
del(df_pand)

### To display the top and bottom of an H2OFrame

In [42]:
import numpy as np

df = h2o.H2OFrame.from_python(np.random.randn(100,4), column_names=list('ABCD'))
# df = h2o.H2OFrame.from_python(np.random.randn(100,4).tolist(), column_names=list('ABCD'))

Parse progress: |█████████████████████████████████████████████████████████| 100%


#### View top 10 rows of the H2OFrame

In [43]:
df.head()

A B C D 
 
 
 -2.31411 -0.420906 -0.784378 -0.600316 
 -0.0303107 -0.334415 0.889986 -0.902756 
 0.889815 0.499498 1.13126 0.457683 
 -0.136728 -0.689726 -1.02643 1.17075 
 -0.152813 -0.649354 -0.580722 0.369689 
 0.0457421 1.50415 -2.58187 0.592722 
 -0.161056 0.251256 1.7856 2.27578 
 0.357449 1.2468 -0.629577 0.788201 
 0.232999 -0.343308 0.824925 -0.46809 
 -0.406646 0.782852 0.605593 0.274735

#### View bottom 5 rows of the H2OFrame

In [44]:
df.tail(5)

A B C D 
 
 
 1.16615 0.959123 0.530487 1.35768 
 -0.554313 0.390748 0.291953 0.00484215 
 0.595438 0.935602 0.883054 -0.068086 
 1.13 -0.858179 -1.14883 2.34633 
 -1.46359 -0.326437 1.39299 0.864175

#### To display the column names

In [45]:
df.columns

['A', 'B', 'C', 'D']

#### To display summary statistics of your data

In [46]:
df.describe()

Rows:100
Cols:4




A B C D 
 
 
 type real real real real 
 mins -2.332030113478489 -2.972182136700504 -2.581868797017038 -2.1546928746485254 
 mean -0.12646894552774224 0.032823221364601536 0.03581393976074404 0.2632227327459989 
 maxs 2.0320337169568012 2.0752372020757757 2.030151665863007 2.894620152252981 
 sigma 0.9858860573408157 0.9662553205729965 1.0069540652319846 1.0370488800291684 
 zeros 0 0 0 0 
 missing 0 0 0 0 
 0 -2.314108666063639 -0.42090562095715506 -0.7843780180590263 -0.6003162897756084 
 1 -0.03031072081031899 -0.33441527804609383 0.8899858455567308 -0.9027563869806008 
 2 0.8898149663554036 0.4994977926679671 1.131262971021223 0.45768320933385154 
 3 -0.13672756234065525 -0.6897257560920579 -1.026433612924891 1.17074910194523 
 4 -0.1528127793972703 -0.6493538387686753 -0.5807220309318316 0.36968934074517273 
 5 0.045742063451174125 1.5041490052106967 -2.581868797017038 0.5927220414956149 
 6 -0.16105622429950334 0.2512559379886862 1.7856023935028105 2.275782161240727 
 7 0.3574492349487698 1.2467958705982851 -0.629576866046083 0.7882008959565284 
 8 0.2329989484126048 -0.34330800244653026 0.8249247771619994 -0.4680898270804136 
 9 -0.40664616830202327 0.7828515920969401 0.6055933849449546 0.27473487701467686

## Selection

### To select a single column by name, resulting in an H2OFrame

In [47]:
df['A']

A 
 
 
 -2.31411 
 -0.0303107 
 0.889815 
 -0.136728 
 -0.152813 
 0.0457421 
 -0.161056 
 0.357449 
 0.232999 
 -0.406646

### To select a single column by index, resulting in an H2OFrame

In [48]:
df[1]

B 
 
 
 -0.420906 
 -0.334415 
 0.499498 
 -0.689726 
 -0.649354 
 1.50415 
 0.251256 
 1.2468 
 -0.343308 
 0.782852

### To select multiple columns by name, resulting in an H2OFrame

In [49]:
df[['B', 'C']]

B C 
 
 
 -0.420906 -0.784378 
 -0.334415 0.889986 
 0.499498 1.13126 
 -0.689726 -1.02643 
 -0.649354 -0.580722 
 1.50415 -2.58187 
 0.251256 1.7856 
 1.2468 -0.629577 
 -0.343308 0.824925 
 0.782852 0.605593

### To select multiple columns by index, resulting in an H2OFrame:

In [50]:
df[0:2]

A B 
 
 
 -2.31411 -0.420906 
 -0.0303107 -0.334415 
 0.889815 0.499498 
 -0.136728 -0.689726 
 -0.152813 -0.649354 
 0.0457421 1.50415 
 -0.161056 0.251256 
 0.357449 1.2468 
 0.232999 -0.343308 
 -0.406646 0.782852

### To select multiple rows by slicing, resulting in an H2OFrame:

__Note__: By default, H2OFrame selection is for columns, so to slice by rows and get all columns, be explicit about selecting all columns:

In [51]:
df[2:7, :]

A B C D 
 
 
 0.889815 0.499498 1.13126 0.457683 
 -0.136728 -0.689726 -1.02643 1.17075 
 -0.152813 -0.649354 -0.580722 0.369689 
 0.0457421 1.50415 -2.58187 0.592722 
 -0.161056 0.251256 1.7856 2.27578

### To select rows based on specic criteria, use Boolean masking

In [52]:
df = h2o.H2OFrame.from_python({'A': [1, 2, 3],
                               'B': ['a', 'a', 'b'],
                               'C': ['hello', 'all', 'world'],
                               'D': ['12MAR2015:11:00:00', '13MAR2015:12:00:00', '14MAR2015:13:00:00']},
                              column_types=['numeric', 'enum', 'string', 'time'])

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [53]:
df[ df["B"] == "a", :]

A B C D 
 
 
 1 a hello 2015-03-12 11:00:00 
 2 a all 2015-03-13 12:00:00

## Missing Data

The H2O parser can handle many different representations of missing data types, including '' (blank), `NA', and None (Python). They are all displayed as __nan__ in Python.

### To create an H2OFrame from Python with missing elements

In [54]:
df = h2o.H2OFrame.from_python(
    {'A': [1, 2, 3, None, ''],
     'B': ['a', 'a', 'b', "NA", 'NA'],
     'C': ['hello', 'all', 'world', None, None],
     'D': ['12MAR2015:11:00:00', None, '13MAR2015:12:00:00', None, '14MAR2015:13:00:00']},
    column_types=['numeric', 'enum', 'string', 'time'])

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [55]:
df

A B C D 
 
 
 1 a hello 2015-03-12 11:00:00 
 2 a all 
 3 b world 2015-03-13 12:00:00 
 nan NA 
 nan NA 2015-03-14 13:00:00

Observation:

    See how None in A, C and D columns are displayed. 
    NA is not treated as Missing    

### To determine which rows are missing data for a given column (`1' indicates missing):

In [56]:
df["A"].isna()

isNA(A) 
 
 
 0 
 0 
 0 
 1 
 1

### To determine the location of all missing data in an H2OFrame

In [57]:
df.isna()

isNA(A) isNA(B) isNA(C) isNA(D) 
 
 
 0 0 0 0 
 0 0 0 1 
 0 0 0 0 
 1 0 0 1 
 1 0 0 0

### To change all missing values in a column to a diferent value:

In [58]:
df[df["A"].isna(), "A"] = 5

In [59]:
df

A B C D 
 
 
 1 a hello 2015-03-12 11:00:00 
 2 a all 
 3 b world 2015-03-13 12:00:00 
 5 NA 
 5 NA 2015-03-14 13:00:00

### Operations

When performing a descriptive statistic on an entire H2OFrame, missing data is generally excluded and the operation is only performed on the columns of the appropriate data type and a native Python object is returned

In [60]:
df = h2o.H2OFrame.from_python(
    {'A': [1, 2, 3, None, '', 6],
     'B': ['a', 'NA', 'c', 'NA', 'NA', 'f'],
     'C': ['hello', 'all', 'world', None, None, 'insofe'],
     'D': ['12MAR2015:11:00:00', None, '13MAR2015:12:00:00', None, '14MAR2015:13:00:00', None]},
    column_types=['numeric', 'enum', 'string', 'time'])

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [61]:
df.mean(na_rm=True)

[3.0, nan, nan, 1426248000000.0]

In [62]:
df["A"].mean(na_rm=False)

[nan]

In [63]:
df["A"].mean(na_rm=True)

[3.0]

In both examples, a native Python object is returned (list and float respectively in these examples).

In [64]:
df.isna()

isNA(A) isNA(B) isNA(C) isNA(D) 
 
 
 0 0 0 0 
 0 0 0 1 
 0 0 0 0 
 1 0 0 1 
 1 0 0 0 
 0 0 0 1

### Handling different types null values
    In H2o null values are recognized as
        1. np.nan 
        2. None in numeric or date columns.
        
    Note: NA values are considered as Strings

In each column, changing the "NA" or "" or None values to np.nan

In [65]:
for cols in df.columns:
    df[df[cols] == "NA", cols] = np.nan
    df[df[cols] == "", cols] = np.nan
    df[df[cols] == None, cols] = np.nan

In [66]:
df

A B C D 
 
 
 1 a hello 2015-03-12 11:00:00 
 2 all 
 3 c world 2015-03-13 12:00:00 
 nan 
 nan 2015-03-14 13:00:00 
 6 f insofe

In [67]:
df.isna()

isNA(A) isNA(B) isNA(C) isNA(D) 
 
 
 0 0 0 0 
 0 1 0 1 
 0 0 0 0 
 1 1 1 1 
 1 1 1 0 
 0 0 0 1

In [68]:
# Filter columns with proportion of NAs greater than ``frac``.
df.filter_na_cols(0.5)

[0.0, 2.0]

In [69]:
df.types

{'A': 'int', 'B': 'enum', 'C': 'string', 'D': 'time'}

### Checking for number null values at column level

In [70]:
df.isna().apply(lambda x: x.sum())

isNA(A) isNA(B) isNA(C) isNA(D) 
 
 
 2 3 2 3

### Handling missing values

#### Imputing the data with central tendencies

In [71]:
# Imputing the numeric columns with mean value
df.impute(column = "A", method = "mean")

[3.0, nan, nan, nan]

In [72]:
# Imputing the categorical columns with mode value
df.impute(column = "B", method = "mode")

[nan, 1.0, nan, nan]

In [73]:
df

A B C D 
 
 
 1 a hello 2015-03-12 11:00:00 
 2 a all 
 3 c world 2015-03-13 12:00:00 
 3 a 
 3 a 2015-03-14 13:00:00 
 6 f insofe

## apply function

When applying functions to each column of the data, an H2OFrame containing the means of each column is returned

In [74]:
df = h2o.H2OFrame.from_python(np.random.randn(100,4), column_names=list('ABCD'))

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [75]:
df

A B C D 
 
 
 -0.156466 -0.0853923 0.953113 0.965269 
 1.3104 0.0622217 0.463746 1.42794 
 0.450265 0.789991 -0.790725 -1.93096 
 -0.280833 -0.662007 0.254932 -0.679072 
 -0.886577 0.393713 1.40827 -1.10921 
 1.55568 -1.38365 0.322107 -0.63462 
 -0.885733 1.79391 0.794088 0.695281 
 0.0638657 -1.62151 -1.20215 0.399627 
 -1.09818 0.0182903 -0.00234107 -1.42411 
 -1.52817 -0.549612 2.42729 0.642553

In [76]:
df.apply(lambda x: x.mean())

A B C D 
 
 
 0.179453 0.0642106 0.148901 -0.011721

When applying functions to each row of the data, an H2OFrame containing the sum of all columns is returned

To determine the number of times a string is contained in each element:

In [77]:
df = h2o.H2OFrame.from_python(['Hello', 'World', 'Welcome', 'To', 'H2O', 'World'])

Parse progress: |█████████████████████████████████████████████████████████| 100%


View the H2OFrame

In [78]:
df

C1 
 
 
 Hello 
 World 
 Welcome 
 To 
 H2O 
 World

Find how many times "l" appears in each string

In [79]:
df.countmatches('l')

C1 
 
 
 2 
 1 
 1 
 0 
 0 
 1

To replace the 1st occurrence of 'l' (lower case letter) with 'x' and return a new H2OFrame

In [80]:
df.sub('l', 'x')

C1 
 
 
 Hexlo 
 Worxd 
 Wexcome 
 To 
 H2O 
 Worxd

For global substitution, use gsub. Both sub and gsub support regular expressions.

In [81]:
df.gsub('l', 'x')

C1 
 
 
 Hexxo 
 Worxd 
 Wexcome 
 To 
 H2O 
 Worxd

To split strings based on a regular expression:

In [82]:
df.strsplit('(l)+')

C1 C2 
 
 
 He o 
 Wor d 
 We come 
 To 
 H2O 
 Wor d

## Merging

In [83]:
# Create df1 frame of random numbers w/ 100 rows
df1 = h2o.H2OFrame.from_python(np.random.randn(100,4).tolist(), column_names=list('ABCD'))

# Create df2 second frame of random numbers w/ 100 rows
df2 = h2o.H2OFrame.from_python(np.random.randn(100,4).tolist(), column_names=list('ABCD'))

Parse progress: |█████████████████████████████████████████████████████████| 100%
Parse progress: |█████████████████████████████████████████████████████████| 100%


Combine the two frames, adding the rows from df2 to df1

In [84]:
df1.rbind(df2)

A B C D 
 
 
 -0.274141 -0.199749 -1.27894 0.998951 
 0.1322 -0.745765 -1.04771 1.16298 
 -1.74891 0.269514 0.0776487 0.353602 
 -1.04062 0.256879 -0.738023 0.634736 
 0.723542 -0.349087 -0.89622 -0.3354 
 -1.27467 1.21818 0.481351 1.09066 
 0.145469 0.733455 0.117387 -0.659801 
 -0.768316 1.04313 0.397842 -0.382767 
 0.790539 0.745677 -0.990822 0.309935 
 -0.805144 0.690823 -0.81115 -0.0128796

In [85]:
df1.rbind(df2).shape

(200, 4)

In [86]:
df1.cbind(df2)

A B C D A0 B0 C0 D0 
 
 
 -0.274141 -0.199749 -1.27894 0.998951 0.101272 0.184185 1.19974 -0.000266256 
 0.1322 -0.745765 -1.04771 1.16298 0.783583 0.316923 1.33438 0.660798 
 -1.74891 0.269514 0.0776487 0.353602 -1.70914 -1.11734 0.513921 0.641114 
 -1.04062 0.256879 -0.738023 0.634736 -1.03772 2.03698 -1.99728 -1.54668 
 0.723542 -0.349087 -0.89622 -0.3354 -0.181565 -1.28046 -1.14866 -0.645552 
 -1.27467 1.21818 0.481351 1.09066 -0.064035 1.1007 2.01732 -1.83818 
 0.145469 0.733455 0.117387 -0.659801 -0.269788 -0.00312191 0.539876 0.794752 
 -0.768316 1.04313 0.397842 -0.382767 -0.229794 -0.519613 -1.04601 1.7119 
 0.790539 0.745677 -0.990822 0.309935 1.3831 1.99276 -0.972583 0.385283 
 -0.805144 0.690823 -0.81115 -0.0128796 0.26339 -1.37165 -0.301477 -0.412487

In [87]:
df1.cbind(df2).shape

(100, 8)

In [88]:
df1 = h2o.H2OFrame.from_python( {'A': ['Hello', 'World', 'Welcome', 'To', 'H2O', 'World'], 
                                 'n': [0, 1, 2, 3, 4, 5]} )
print(df1)

Parse progress: |█████████████████████████████████████████████████████████| 100%


A n 
 
 
 Hello 0 
 World 1 
 Welcome 2 
 To 3 
 H2O 4 
 World 5

In [89]:
# Create a single-column, 100-row frame
# Include random integers from 0-5
df2 = h2o.H2OFrame.from_python(np.random.randint(0, 6, (100,1)), column_names= list('n'))
print(df2)

Parse progress: |█████████████████████████████████████████████████████████| 100%


n 
 
 
 3 
 0 
 3 
 4 
 4 
 3 
 3 
 5 
 3 
 1

In [90]:
# Combine column "n" from both datasets
print(df2.merge(df1))
print(df2.merge(df1).shape)

n A 
 
 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello 
 0 Hello


(100, 2)


## Grouping

"Grouping" refers to the following process:

* splitting the data into groups based on some criteria
* applying a function to each group independently
* combining the results into an H2OFrame

In [91]:
df = h2o.H2OFrame({'A' : ['foo', 'bar', 'foo', 'bar', 'foo', 'bar', 'foo', 'foo'],
                   'B' : ['one', 'one', 'two', 'three', 'two', 'two', 'one', 'three'],
                   'C' : np.random.randn(8).tolist(),
                   'D' : np.random.randn(8).tolist()})

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [92]:
df

A B C D 
 
 
 foo one -0.306573 -0.0850366 
 bar one -2.17191 -0.675158 
 foo two -1.30353 0.150156 
 bar three -0.573381 0.74003 
 foo two -0.537055 0.050292 
 bar two 0.29252 0.151292 
 foo one -1.64412 -0.499572 
 foo three 0.314555 0.825306

To group and then apply a function to the results

In [93]:
df.group_by('A').sum().frame

A sum_B sum_C sum_D 
 
 
 bar 3 -2.45277 0.216163 
 foo 5 -3.47673 0.441146

To group by multiple columns and then apply a function:

In [94]:
df.group_by(['A','B']).sum().frame

A B sum_C sum_D 
 
 
 bar one -2.17191 -0.675158 
 bar three -0.573381 0.74003 
 bar two 0.29252 0.151292 
 foo one -1.95069 -0.584609 
 foo three 0.314555 0.825306 
 foo two -1.84059 0.200448

## Using Date and Time Data

In [95]:
df = h2o.H2OFrame.from_python({'D': ['18OCT2015:11:00:00', '19OCT2015:12:00:00', '20OCT2015:13:00:00']}, 
                              column_types=['time'])

df

Parse progress: |█████████████████████████████████████████████████████████| 100%


D 
 
 
 2015-10-18 11:00:00 
 2015-10-19 12:00:00 
 2015-10-20 13:00:00

In [96]:
df.types

{'D': 'time'}

To display the day of the month:

In [97]:
df['D'].day()

D 
 
 
 18 
 19 
 20

To display the day of the week

In [98]:
df['D'].dayOfWeek()

D 
 
 
 Sun 
 Mon 
 Tue

## Categoricals

    H2O handles categorical (also known as enumerated or factor) values in an H2OFrame. This is signicant because categorical columns have specific treatments in each of the machine learning algorithms 

In [99]:
df = h2o.H2OFrame({'A' : ['foo', 'bar', 'foo', 'bar', 'foo', 'bar', 'foo', 'foo'],
                   'B' : ['one', 'one', 'two', 'three', 'two', 'two', 'one', 'three'],
                   'C' : np.random.randn(8).tolist(),
                   'D' : np.random.randn(8).tolist()})

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [100]:
df.types

{'A': 'enum', 'B': 'enum', 'C': 'real', 'D': 'real'}

To determine if any column is a categorical/enumerated/factor type:

In [101]:
df.anyfactor()

True

To view the categorical levels in a single column:

In [102]:
df["A"].levels()

[['bar', 'foo']]

In [103]:
h2o.remove_all()

## Loading and Saving Data

In addition to loading data from Python objects, H2O can load data directly from:

* disk
* network file systems (NFS, S3)
* distributed file systems (HDFS)
* HTTP addresses

H2O currently supports the following file types:
    
* CSV (delimited) files
* ORC
* SVMLite
* Parquet
* ARFF
* XLS
* XLSX
* AVRO

### To load data from the same machine running H2O:

### To load data from the machine(s) running H2O to the machine running Python:

### To save an H2OFrame on the machine running H2O:

### To save an H2OFrame on the machine running Python:

#### upload_file

In [104]:
localFile = "data/iris.csv"

#### Cannot load the data from extrenal path. Filesneeds to be present in local machine

In [105]:
# Reading files from local path or machine
df1 = h2o.upload_file(localFile, destination_frame="uploadfile.hex")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [106]:
print(type(df1))
print(df1.columns)
print(df1.shape)

<class 'h2o.frame.H2OFrame'>
['C1', 'C2', 'C3', 'C4', 'C5']
(150, 5)


In [107]:
# Reading files from remote or external location
df2 = h2o.upload_file(localFile, destination_frame="uploadfile.hex",
                      col_names=['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width', 'Species'])

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [108]:
h2o.ls()

key
0  uploadfile.hex

In [109]:
h2o.download_csv(df2, "irisdownloadfile.csv")

'/home/jeevan/Desktop/H2o_Python/irisdownloadfile.csv'

In [110]:
h2o.ls()

key
0  uploadfile.hex

In [111]:
h2o.remove_all()

## Shutdwon the h2o cluster

In [112]:
h2o.cluster().shutdown()

H2O session _sid_a718 closed.
